### DISCUSSION TITLE
How to check code with short opening tags?

### DISCUSSION THREAD

**Discussion by adam-nielsen** [PHPCSStandards/PHP_CodeSniffer#415] [Q&A] [answered]
I have code that uses short PHP tags, like this:

```php
<?php

function example() {
echo "Example";
?><span><a href="<?= 'test' ?>>Example</a></span><?
 echo "More PHP code";
}
```

However everything after a short opening tag gets identified as HTML:

```
Process token 17 on line 5 [col:18;len:3;lvl:0;]: T_OPEN_TAG_WITH_ECHO => <?=
        Process token 18 on line 5 [col:21;len:1;lvl:0;]: T_WHITESPACE => ·
        Process token 19 on line 5 [col:22;len:6;lvl:0;]: T_CONSTANT_ENCAPSED_STRING => 'test'
        Process token 20 on line 5 [col:28;len:1;lvl:0;]: T_WHITESPACE => ·
        Process token 21 on line 5 [col:29;len:2;lvl:0;]: T_CLOSE_TAG => ?>
        Process token 22 on line 5 [col:31;len:21;lvl:0;]: T_INLINE_HTML => >Example</a></span><?\n
        Process token 23 on line 6 [col:1;len:22;lvl:0;]: T_INLINE_HTML => ·echo·"More·PHP·code";\n
        Process token 24 on line 7 [col:1;len:1;lvl:0;]: T_INLINE_HTML => }\n
```

While the above code parses correctly, my codebase ends up crashing phpcbf because of this issue:

```
PHP Fatal error:  Uncaught PHP_CodeSniffer\Exceptions\RuntimeException: Undefined array key -1 in phar://phpcbf.phar/src/Standards/PSR12/Sniffs/Functions/ReturnTypeDeclarationSniff.php on line 3 in phar://phpcbf.phar/src/Runner.php:3
Stack trace:
#0 phar://phpcbf.phar/src/Standards/PSR12/Sniffs/Functions/ReturnTypeDeclarationSniff.php(3): PHP_CodeSniffer\Runner->handleErrors()
#1 phar://phpcbf.phar/src/Files/File.php(3): PHP_CodeSniffer\Standards\PSR12\Sniffs\Functions\ReturnTypeDeclarationSniff->process()
#2 phar://phpcbf.phar/src/Files/LocalFile.php(3): PHP_CodeSniffer\Files\File->process()
#3 phar://phpcbf.phar/src/Fixer.php(3): PHP_CodeSniffer\Files\LocalFile->process()
#4 phar://phpcbf.phar/src/Reports/Cbf.php(3): PHP_CodeSniffer\Fixer->fixFile()
#5 phar://phpcbf.phar/src/Reporter.php(3): PHP_CodeSniffer\Reports\Cbf->generateFileReport()
#6 phar://phpcbf.phar/src/Runner.php(3): PHP_CodeSniffer\Reporter->cacheFileReport()
#7 phar://phpcbf.phar/src/Runner.php(3): PHP_CodeSniffer\Runner->processFile()
#8 phar://phpcbf.phar/src/Runner.php(3): PHP_CodeSniffer\Runner->run()
#9 phpcbf.phar(6): PHP_CodeSniffer\Runner->runPHPCBF()
#10 {main}
  thrown in phar://phpcbf.phar/src/Runner.php on line 3
```

I tried to enable short tags by putting this in my ruleset XML file:

```xml
  <ini name="short_open_tag" value="On"/>
```

However it made no difference.  How can I get phpcbf to correctly handle short opening tags without changing my global PHP settings?

(I don't want to change my global PHP settings because I don't want to enable short tags for my own code, but I am using pbpcbf to check someone else's legacy codebase.)

**Comment by jrfnl** [comment]
> However everything after a short opening tag gets identified as HTML

That's not a PHPCS issue, but a characteristic of how PHP parses the code when `short_open_tags=Off`.

> How can I get phpcbf to correctly handle short opening tags without changing my global PHP settings?

You can't.

As per the [short_open_tags documentation](https://www.php.net/manual/en/ini.core.php#ini.sect.language-options), the value of `short_open_tags` cannot be [changed at runtime](https://www.php.net/manual/en/configuration.changes.modes.php) as it is a `INI_PERDIR` type setting.

Not sure if changing the short open tags to long tags was on the menu for cleaning up the codebase too, but if so, the `Generic.PHP.DisallowShortOpenTag` sniff should be able to help you find most uses of short open tags in the code base, even with `short_open_tags=Off` - as long as the open tag is accompanied by a PHP close tag (in other words: it would not flag the use in the above code sample as there is no close tag at the end of the file).

Once the short tags have been fixed to long tags, everything else should start working properly.

**Reply by jrfnl** [reply, marked as answer]
Also see: #416 ;-)

**Reply by adam-nielsen** [reply]
Interesting, thanks for the info.  I didn't realise it was PHP itself parsing the code.  I managed to get it to work with `-d` as hinted to in #416:

    php -d short_open_tag=on phpcbf.phar ...

Caught me out at first because I had `-d` after the `phar` file, but it has to come before.  So at least that allows me to avoid modifying the global .ini file.

I don't really want to change the short tags to long tags because there's so much PHP embedded within HTML it will make it even less readable than it already is!  So just going to leave the short tags as-is for now.

**Reply by fredden** [reply]
Note that the short open echo cannot be disabled, so `<?= $value ?>` will always work; it's just the `<?` that gets turned on/off with that setting.

**Reply by jrfnl** [reply]
@adam-nielsen Glad to hear you got things working!

> Caught me out at first because I had -d after the phar file, but it has to come before. So at least that allows me to avoid modifying the global .ini file.

Interesting... I actually tested that locally yesterday while writing my response and I didn't get that working (not even when using `php -d ... phpcs.phar ...`), so happy to hear it worked for you.

**Reply by adam-nielsen** [reply]
@jrfnl Interesting!  This was on Linux with PHP 8.3.3 in case that makes a difference.  It definitely works with `-d`, in so much as it gets further than before and comes up with entirely new errors 😆

**Comment by jrfnl** [comment]
FYI: While it doesn't directly address this issue, PR #447 is a proposal to improve the help screens, including improved information on the help screen about what PHP ini options are changeable via PHPCS itself.

Reviews and testing of that PR would be appreciated.
